storage/locks.go

```diff
@@ -0,0 +1,56 @@
+package storage
+
+import (
+	"sync"
+
+	"github.com/jordanschalm/lockctx"
+)
+
+// This file enumerates all named locks used by the storage layer.
+
+const (
+	// LockInsertBlock protects the entire block insertion process (Extend or ExtendCertified)
+	LockInsertBlock = "lock_insert_block"
+	// LockFinalizeBlock protects the entire block finalization process (Finalize)
+	LockFinalizeBlock = "lock_finalize_block"
+)
+
+// Locks returns a list of all named locks used by the storage layer.
+func Locks() []string {
+	return []string{LockInsertBlock, LockFinalizeBlock}
+}
+
+// makeLockPolicy constructs the policy used by the storage layer to prevent deadlocks.
+// We use a policy defined by a directed acyclic graph, where nodes are named locks.
+// A directed edge between two nodes A, B means: I can acquire B next after acquiring A.
+// When no edges are added, each lock context may acquire at most one lock.
+// This function will panic if a policy is created which does not prevent deadlocks.
+func makeLockPolicy() lockctx.Policy {
+	return lockctx.NewDAGPolicyBuilder().
+		Add(LockInsertBlock, LockFinalizeBlock).
+		Build()
+}
+
+var makeLockManagerOnce sync.Once
+
+// MakeSingletonLockManager returns the lock manager used by the storage layer.
+// This function must be used for production builds and must be called exactly once process-wide.
+// If this function is called more than once, it will panic. This is strictly enforced because
```

Comment:
This is not the typical singleton pattern, which can be called multiple time. Why not allowing it to be called multiple times, and always return the same global singleton?

Reply:
The reason I did it this way is because our existing structure uses dependency injection and avoids typical singleton getters which can be called multiple times. My goal with this function is to:
- continue using the dependency injection pattern (the `LockManager` dependency should be constructed once then passed down to components which depend on it)
- but also, fail loudly if someone misuses this function by invoking it multiple times